subsys/pm/pm_stats.c

```diff
@@ -32,44 +25,48 @@ STATS_NAME(pm_cpu_stats, state_last_cycles)
 STATS_NAME(pm_cpu_stats, state_total_cycles)
 STATS_NAME_END(pm_cpu_stats);
 
+static STATS_SECT_DECL(pm_cpu_stats) stats[CONFIG_MP_NUM_CPUS][PM_STATE_COUNT];
+
 #define PM_STAT_NAME_LEN sizeof("pm_cpu_XXX_state_X_stats")
-static char pm_cpu_stat_names[CONFIG_MP_NUM_CPUS][PM_STATE_COUNT][PM_STAT_NAME_LEN];
-static struct stats_pm_cpu_stats pm_cpu_stats[CONFIG_MP_NUM_CPUS][PM_STATE_COUNT];
+static char names[CONFIG_MP_NUM_CPUS][PM_STATE_COUNT][PM_STAT_NAME_LEN];
+static uint32_t time_start[CONFIG_MP_NUM_CPUS];
+static uint32_t time_stop[CONFIG_MP_NUM_CPUS];
 
-static int pm_stats_init(const struct device *unused)
+static int pm_stats_init(const struct device *dev)
 {
-	for (int i = 0; i < CONFIG_MP_NUM_CPUS; i++) {
-		for (int j = 0; j < PM_STATE_COUNT; j++) {
-			snprintk(pm_cpu_stat_names[i][j], PM_STAT_NAME_LEN,
+	ARG_UNUSED(dev);
+
+	for (uint8_t i = 0U; i < CONFIG_MP_NUM_CPUS; i++) {
+		for (uint8_t j = 0U; j < PM_STATE_COUNT; j++) {
+			snprintk(names[i][j], PM_STAT_NAME_LEN,
```

Comment:
Good idea, let me see if I can do it

Reply:
I tried this before but ran into something... can't recall what now though

Reply:
This is the best macrobatics I've come up with (UTIL_LISTIFY can't be nested) (note that naming is re-arranged so that state comes before cpu index):
```
#define NAME_INIT(state_idx) "pm_state_" STRINGIFY(state_idx)
#define NAMES_INIT(cpu_idx, _)						       \
	{ FOR_EACH(NAME_INIT, ("_cpu_" STRINGIFY(cpu_idx) "_stats",),	       \
		   0, 1, 2, 3, 4, 5, 6) "_cpu_" STRINGIFY(cpu_idx) "_stats" },
static const char *names[CONFIG_MP_NUM_CPUS][PM_STATE_COUNT] = {
	UTIL_LISTIFY(CONFIG_MP_NUM_CPUS, NAMES_INIT)
};
```

not sure if it's worth sacrificing readability for something that is usually enabled to debug/measure performance.

Reply:
Thanks for trying.  I agree the readability is better with the for loops.